pagefaultgames/pokerogue
# [Move] Fully implement Rage

## What are the changes the user will see?
Rage now correctly applies an attack boost when a Pokémon is hit and Rage is the last move they used. 

Some important notes about Rage's intended behavior:
- A pokemon **starts** raging and gains the `RageTag` tag if it successfully hits a non-immune target with Rage.
- A pokemon **stops** raging and loses the `RageTag` tag if it successfully uses a move that is not Rage, even if that move misses and/or fails.
- A pokemon does **NOT** stop raging if
  - It uses rage and misses or the target is immune. 
  - It was commanded to use a different move but has not acted yet
  - The Run or Ball options are selected in battle
  - It attempts to use a different move but is unable to act (via being flinched, asleep, confused, etc) 

## Why am I making these changes?
Rage is not fully implemented #3503 

## What are the changes from a developer perspective?
- `data/battler-tags`: Added new `RageTag` to mark that the holder is raging. Uses `MOVE_EFFECT` and `CUSTOM` lapse types, to check if the tag should be removed because the holder used a different move, and to apply the attack boost when hit respectively. 
- `phases/move-effect-phase`: Added code to lapse a target's `RageTag`when they are hit, in line with the implementations of Shell Trap and Beak Blast.
- `data/move`: Added `AddBattlerTagAttr` to Rage to apply a `RageTag` to the user on a successful hit.
- `test/moves/rage.test`:  New test file to test that rage gives attack boosts as intended
- `locales`: Rage uses a custom message for its attack boosts, and since there's no other indicator I also added a separate message that doesn't exist in game for when a Pokémon successfully enters a rage. All non-English text is placeholder, and is just a copy of the English text.

(Added to the main post from a comment so people don't miss it):
The  "rage is starting to build" text is not present in the mainline games, I added it tentatively in part because otherwise there is no in-game indicator that a pokemon has successfully started raging, and mostly because I needed to test things and record example videos and it saved me from having to play through an extra turn every time to check if the tag was/wasn't applied correctly.

It's a slight deviation from cartridge, but since Rage is a bit unintuitive, I figured I'd leave it in until people wanted it gone.

### Screenshots/Videos
A raging pokemon gains an attack boost when hit:
A pokemon stays raging between turns:
A pokemon does not stop raging until it uses a non-rage move:

https://github.com/user-attachments/assets/92f4a7fb-5793-4978-9b9f-dc65effa3a47

A pokemon does not start raging until it hits a target with rage:

https://github.com/user-attachments/assets/8240684b-1fd8-4cd7-adef-4c195eb74415

A pokemon does not start raging if the target is immune:

https://github.com/user-attachments/assets/f27bd885-bb02-41e4-b417-7c714a96cc65

A Pokémon does not start raging if the initial Rage misses:

https://github.com/user-attachments/assets/0c9f2999-a657-475c-aef6-1dcde474b7b0

A Pokémon does not stop raging if the Player selects Run or Ball:

https://github.com/user-attachments/assets/e021103f-9da1-4962-869c-4d872787425e

Rage provides an attack boost per instance of damage:

https://github.com/user-attachments/assets/f74927a0-f50d-4a67-a0ae-1a45a626142a

## How to test the changes?
There are unit tests written in `tests/moves/rage.test.ts` that should hopefully cover most functionality! Currently I am aware that there are not tests for: 
- When the player selects Run or Ball
- When Rage just misses normally
- Comprehensive checks for any discrepancy in behavior between Player and Enemy Pokemon. (Most tests cover one or the other, but rarely both)
In addition, the testing for when Rage fails or has no effect includes specific tests against Ghost types and against Pokemon who are semi-invulnerable through the use of Fly, but may not be exhaustive. 

For the tests, just make sure that they look right and are running without issue, and potentially double check them using overrides, otherwise, you can check for any edge cases I may have missed also using the options in `overrides.ts`. I described Rage's expected behavior further up, and it can be cross referenced [here](https://bulbapedia.bulbagarden.net/wiki/Rage_(move))

## Checklist
